src/clp_ffi_js/ir/StructuredIrStreamReader.cpp

```diff
@@ -139,12 +145,58 @@ auto StructuredIrStreamReader::get_filtered_log_event_map() const -> FilteredLog
     return FilteredLogEventMapTsType{emscripten::val::array(m_filtered_log_event_map.value())};
 }
 
-void StructuredIrStreamReader::filter_log_events(LogLevelFilterTsType const& log_level_filter) {
-    generic_filter_log_events(
-            m_filtered_log_event_map,
-            log_level_filter,
-            *m_deserialized_log_events
-    );
+void StructuredIrStreamReader::filter_log_events(
+        LogLevelFilterTsType const& log_level_filter,
+        std::string const& kql_filter
+) {
+    m_filtered_log_event_map = std::nullopt;
+
+    if (false == kql_filter.empty()) {
+        auto& reader{m_stream_reader_data_context->get_reader()};
+        reader.seek_from_begin(0);
+        auto indices{query_log_event_indices(reader, kql_filter)};
+        m_filtered_log_event_map = std::move(indices);
+    }
+
+    if (false == log_level_filter.isNull()) {
+        std::vector<size_t> filtered_log_event_map;
+
+        auto const filter_levels{
+                emscripten::vecFromJSArray<std::underlying_type_t<LogLevel>>(log_level_filter)
+        };
+
+        auto filter_and_collect_idx = [&](size_t const log_event_idx) {
+            auto const& log_event{m_deserialized_log_events->at(log_event_idx)};
+            if (std::ranges::find(
+                        filter_levels,
+                        clp::enum_to_underlying_type(log_event.get_log_level())
+                )
+                != filter_levels.end())
+            {
+                filtered_log_event_map.emplace_back(log_event_idx);
+            }
+        };
+
+        if (m_filtered_log_event_map.has_value()) {
+            for (auto const log_event_idx : m_filtered_log_event_map.value()) {
+                filter_and_collect_idx(log_event_idx);
+            }
+        } else {
+            for (size_t log_event_idx = 0; log_event_idx < m_deserialized_log_events->size();
```

Comment:
```suggestion
            for (size_t log_event_idx{0}; log_event_idx < m_deserialized_log_events->size();
```
Nit: For consistency.